[Rollup] Use composite's missing_bucket

polyfractal opened the pull request:
We can leverage the composite agg's new `missing_bucket` feature on terms groupings.  This means the aggregation criteria used in the indexer will now return null buckets for missing keys.  We then index these and rely on a default `null_value` on the Rollup's mapping to inject a placeholder.

On the search side of the house, we can remove the placeholder when unrolling the response.

By indexing null values, we can guarantee correct doc counts with "combined" jobs (where a job rolls up multiple schemas).  This was previously impossible since composite would ignore documents that didn't have _all_ the keys, meaning non-overlapping schemas would cause composite to return no buckets.

The documentation has been adjusted to recommend a single, combined job.  It also makes reference to the previous issue to help users that are upgrading (rather than just deleting the sections).

**BWC**
This change is somewhat incompatible with older rollup jobs, because all jobs in a shared index share the same mapping templates.  So if the template was updated, new jobs would be "fixed" but old jobs would start getting `null_values` and end up in an inconsistent state.  But not updating the index meant the new jobs would also be "broken".

I opted to just disallow new jobs from being created rollup indices that were pre-6.4.0.  So the user will have to create a new rollup index for new jobs.

**Important Note:**
This PR enables `missing_bucket` on all groups (terms, histo, date_histo), but we only enable a `null_value` on string terms.  There's not a reasonable default missing value for numerics as all numerics are technically a real value.  I want to follow this PR up with another that allows the user to configure a missing value for histo/date_histo if they desire. 

Otherwise, the null values will just be indexed as null and dropped as usual, so this only "fixes" strings.  But that's the major problem at the moment.

elasticmachine commented:
Pinging @elastic/es-search-aggs

colings86 commented on x-pack/docs/en/rollup/understanding-groups.asciidoc:
```diff
@@ -121,16 +121,15 @@ if a field is useful for aggregating later, and how you might wish to use it (te
-_all_ of the keys `A, B` and `C`.
+There was previously a limitation in how Rollup could handle indices that had heterogeneous mappings (multiple, unrelated/non-overlapping
+mappings).  The recommendation at the time was to configure a separate job per data "type".  For example, you might configure a separate
+job for each Beat module that you had enabled (one for `process`, another for `filesystem`, etc).
```
nit: I think it should be `Beats module`?

colings86 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/Rollup.java:
```diff
@@ -85,6 +85,8 @@
     public static final String ROLLUP_TEMPLATE_VERSION_PATTERN =
             Pattern.quote("${rollup.dynamic_template.version}");
+    public static final String ROLLUP_NULL_VALUE_PLACEHOLDER = "ROLLUP_NULL_VALUE_PLACEHOLDER";
+    public static final String ROLLUP_NULL_VALUE = "__ROLLUP_NULL_VALUE_PLACEHOLDER__";
```
These look like they might be for the same purpose, maybe we should add comments to them to explain why they are both needed and what for?

colings86 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupResponseTranslator.java:
```diff
@@ -382,6 +382,12 @@ private static InternalAggregation unrollMultiBucket(InternalMultiBucketAggregat
+                // Hide our `null_value` placeholder so it doesn't show up in the terms list
+                if (bucket.getKeyAsString().equals(Rollup.ROLLUP_NULL_VALUE)) {
+                    return null;
+                }
```
Why is this not needed for LongTerms and the histograms? do they not have the concept of a null bucket as well?

colings86 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java:
```diff
@@ -176,6 +178,21 @@ static void updateMapping(RollupJob job, ActionListener<PutRollupJobAction.Respo
+
+            String stringVersion = (String)((Map<String, Object>) m).get(Rollup.ROLLUP_TEMPLATE_VERSION_FIELD);
+            if (stringVersion == null) {
+                logger.warn("Could not determine version of existing rollup metadata for index [" + indexName + "]");
```
Is a warning enough here? I am wondering if when the stringVersion is null: a) will we end up throwing a NPE below such as in the `Version.fromString()` method and b) if we will be able to "do the right thing"?

colings86 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java:
```diff
@@ -176,6 +178,21 @@ static void updateMapping(RollupJob job, ActionListener<PutRollupJobAction.Respo
+            if (parsedVersion.before(Version.V_6_4_0)) {
+                String msg = "Cannot create rollup job [" + job.getConfig().getId() + "] because the rollup index contains " +
+                    "jobs from pre-6.4.0.  The mappings for these jobs are not compatible with 6.4.0+.  Please specify a new rollup " +
+                    "index.";
```
Does this mean that users are going to need to recreate their existing rollup jobs and rollup indexes when they upgrade from 6.3 to 6.4+?

colings86 commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/job/IndexerUtils.java:
```diff
@@ -90,7 +90,10 @@ private static CRC32 processKeys(Map<String, Object> keys, Map<String, Object> d
-                if (v instanceof String) {
+                if (v == null) {
+                    // Arbitrary value to update the doc ID with for nulls
+                    docID.update(19);
```
disappointed you didn't make this value `42` 😃

polyfractal commented:
Hmm hmm... I think you're right.  I was assuming we had to represent those nulls for accurate counts, but we don't care if the field is null so long as the _rest_ of the fields are present (whereas before the entire document was omitted if one of the keys was gone).

Let me think about it, but I think you're right.  That also means we don't have the compatibility issue anymore if we don't have to set a `null_value`

polyfractal commented:
Ok, so I just pushed a commit which:

- Removed `null_value`
- Removed the rollup version compatibility check (since it was the use of `null_value` which made it incompatible)
- Removed the full cluster restart test, since there's not incompatibility to check anymore
- Added some logic to handle null numeric/histo keys which I missed before... those would have NPEd.  Added a test to verify that behavior

I left in the check for rollup version in general (since that's a useful validation) as well as the tweaks to the full cluster restart helper methods, since that will make future tests easier.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/DateHistoGroupConfig.java:
```diff
@@ -159,7 +159,7 @@ public Rounding createRounding() {
         vsBuilder.field(field);
         vsBuilder.timeZone(timeZone);
-
+        vsBuilder.missingBucket(true);
```
We shouldn't allow missing bucket here. We use this config for the timestamp field which a required field for rollup documents. Moreover since it's the first field in the composite, enabling missing buckets would disable the composite optimization when the field is indexed. IMO it's ok to require the timestamp field to be present since otherwise we have no way to know if the document should be part of the rollup.

polyfractal replied:
Ah yeah, silly me.  Absolutely shouldn't allow missing on the timestamp field :)

Good catch thanks

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/job/IndexerUtils.java:
```diff
@@ -83,14 +83,27 @@ private static CRC32 processKeys(Map<String, Object> keys, Map<String, Object> d
                 doc.put(k  + "." + RollupField.INTERVAL, groupConfig.getDateHisto().getInterval());
                 doc.put(k  + "." + DateHistoGroupConfig.TIME_ZONE, groupConfig.getDateHisto().getTimeZone().toString());
-                docID.update(Numbers.longToBytes((Long)v), 0, 8);
+                if (v == null) {
```
we should not allow null here

jimczi commented on x-pack/docs/en/rollup/understanding-groups.asciidoc:
```diff
@@ -191,222 +190,14 @@ PUT _xpack/rollup/job/combined
-when this particular scenario is fixed.
+The rollup job will automatically use a placeholder term (`__ROLLUP_NULL_VALUE_PLACEHOLDER__`) as the `null_value` for keyword fields,
+which allows it to handle documents that may be missing some of the grouping fields.  This placeholder is then removed from search
+results, resulting in correct doc counts in a manner that is invisible to the user.
```
This part should be updated since we don't use the placeholder term anymore.

polyfractal commented:
Review comments addressed, tests back to passing too.

polyfractal commented:
Woo, thanks @jimczi!  And good catch on the `null_value`, I'm happy we didn't need to do that. :)
